src/main/kotlin/com/exactpro/th2/conn/dirty/tcp/core/Microservice.kt

```diff
@@ -0,0 +1,248 @@
+        }.onFailure {
+            throw IllegalStateException("Failed to subscribe to input queue", it)
+        }
+
+        if (settings.autoStart) manager.openAll(settings.autoStopAfter)
+    }
+
+    private fun handleBatch(tag: String, batch: MessageGroupBatch) {
+        batch.groupsList.forEach { group ->
+            group.runCatching(::handleGroup).recoverCatching { cause ->
+                onError("Failed to handle message group", group, cause)
+            }
+        }
+    }
+
+    private fun handleGroup(group: MessageGroup) {
+        if (group.messagesCount != 1) {
+            onError("Message group must contain only a single message", group)
+            return
+        }
+
+        val message = group.messagesList[0]
+
+        if (!message.hasRawMessage()) {
+            onError("Message is not a raw message", message)
+            return
+        }
+
+        val rawMessage = message.rawMessage
+        val sessionAlias = rawMessage.sessionAlias
+
+        if (sessionAlias !in channels) {
+            onError("Unknown session alias: $sessionAlias", message)
+            return
+        }
+
+        actionExecutor.execute("send-$sessionAlias") {
+            val client = manager.open(sessionAlias, settings.autoStopAfter)
+
+            rawMessage.runCatching(client::send).recoverCatching { cause ->
+                onError("Failed to send message", message, cause)
+            }
+        }
+    }
+
+    private fun createChannel(session: SessionSettings): Channel {
+        val parentEventId = eventRouter.storeEvent(session.sessionAlias.toEvent(), rootEventId).id
+        val sendEvent: (Event) -> Unit = { eventRouter.storeEvent(it, parentEventId) }
+
+        val handlerContext = Context(session.handler, readDictionary, sendEvent)
+        val handler = handlerFactory.create(handlerContext).apply {
+            registerResource("handler-${session.sessionAlias}", ::close)
+        }
+
+        val manglerContext = Context(session.mangler, readDictionary, sendEvent)
+        val mangler = manglerFactory.create(manglerContext).apply {
+            registerResource("mangler-${session.sessionAlias}", ::close)
+        }
+
+        val channel = Channel(
+            InetSocketAddress(session.host, session.port),
+            session.secure,
+            session.sessionAlias,
+            settings.reconnectDelay,
+            handler,
+            mangler,
+            ::onEvent,
+            ::onMessage,
+            eventLoopGroup,
+            actionExecutor,
+            EventID.newBuilder().setId(parentEventId).build()
+        )
+
+        handlerContext.channel = channel
+        manglerContext.channel = channel
+
+        return channel
+    }
+
+    private fun publishSentEvents(batch: MessageGroupBatch) {
+        if (!settings.publishSentEvents) return
+
+        val messageIds = batch.groupsList.groupBy(
+            { it.messagesList[0].parentEventId!! },
+            { it.messagesList[0].messageId }
```

Comment:
Some of the messages can miss parentEventId, for example, handler's or mangler's messages. Conn should emit events for messages are income from other th2 boxes, for example codec.